nats.go

```diff
@@ -1542,6 +1542,8 @@ func (s *Subscription) NextMsg(timeout time.Duration) (msg *Msg, err error) {
 			return nil, ErrConnectionClosed
 		}
 		delivered := atomic.AddUint64(&s.delivered, 1)
+		// FIXME: jam: off to the races on max vs. delivered
+		// FIXME: jam: cached above but possibly updated...
 		if max > 0 {
 			if delivered > max {
 				return nil, errors.New("nats: Max messages delivered")
```

Comment:
If you don't check delivered here, and you propose to remove from processMsg() then where is it going to happen? 
In case of sync subscriptions (use of NextMsg), the user can call it from different places at the same time. The use of atomic guarantees that only the first caller to go over the max will remove the subscription. The other calls to NextMsg() would have delivered strictly greater than max and will simply return the exception.